src/Elcodi/Admin/ProductBundle/Form/Type/CategoryType.php

```diff
@@ -82,51 +93,60 @@ public function setDefaultOptions(OptionsResolverInterface $resolver)
      */
     public function buildForm(FormBuilderInterface $builder, array $options)
     {
+        $currentCategoryId = $builder->getData()->getId();
+        $parentEntityWhereClause = $currentCategoryId
+            ? "c.id <> $currentCategoryId AND c.root = 1"
+            : "c.root = 1";
+
         $builder
-            ->add('name', 'text', array(
+            ->add('name', 'text', [
                 'required' => true,
                 'label'    => 'name',
-            ))
-            ->add('slug', 'text', array(
+            ])
+            ->add('slug', 'text', [
                 'required' => false,
                 'label'    => 'slug',
-            ))
-            ->add('root', 'checkbox', array(
+            ])
+            ->add('root', 'checkbox', [
                 'required' => false,
                 'label'    => 'Root Category',
-            ))
-            ->add('position', 'integer', array(
+            ])
+            ->add('position', 'integer', [
                 'required' => false,
                 'label'    => 'Position',
-            ))
-            ->add('enabled', 'checkbox', array(
+            ])
+            ->add('enabled', 'checkbox', [
                 'required' => false,
                 'label'    => 'Visible',
-            ))
-            ->add('metaTitle', 'text', array(
+            ])
+            ->add('metaTitle', 'text', [
                 'required' => false,
                 'label'    => 'metaTitle',
-            ))
-            ->add('metaDescription', 'text', array(
+            ])
+            ->add('metaDescription', 'text', [
                 'required' => false,
                 'label'    => 'metaDescription',
-            ))
-            ->add('metaKeywords', 'text', array(
+            ])
+            ->add('metaKeywords', 'text', [
                 'required' => false,
                 'label'    => 'metaKeywords',
-            ))
-            ->add('parent', 'entity', array(
-                'class'    => $this->categoryFactory->getEntityNamespace(),
-                'required' => false,
-                'label'    => 'parent',
-                'multiple' => false,
-            ))
-            ->add('products', 'entity', array(
+            ])
+            ->add('parent', 'entity', [
+                'class'         => $this->categoryFactory->getEntityNamespace(),
+                'query_builder' => $this
```

Comment:
I prefer the lazy approach using a closure. This removes also the dependency on the repository, since [it is passed as a parameter to the function](http://symfony.com/doc/current/reference/forms/types/entity.html#query-builder).